_source/_posts/2022-03-15-programming-security-and-why-rust.md

```diff
@@ -0,0 +1,157 @@
+
+# Why Rust?
+
+Now that we understand how important programming safety is, let's see why Rust is one of the safest languages and how it can avoid most of the security issues we normally encounter with languages like C/C++.
+
+For those not familiar, Rust is a high-level multi-paradigm language. It's ideal for functional and imperative programming. It has very modern and, in my opinion, the best tooling for a programming language. Though it was originally designed as a systems programming language, its advantages and flexibility have made it suitable for all sorts of use cases as a general-purpose language.
+
+> "Rust throws around some buzz words in its docs, but they are not just marketing buzz; they actually mean it with full sincerity, and they matter a lot."
+
+## Safety guarantee
+
+{% img blog/programming-security-and-why-rust/safety-meme.jpg alt:"memory safety meme" width:"600" %}{: .center-image }
+
+The safety guarantee is one of the most important aspects of Rust; Rust is memory-safe, null-safe, type-safe, and thread-safe by design. If the compiler detects unsafe code, it will refuse to compile that code by default. You would have to go out of your way to break those guarantees using the `unsafe` keyword. So even in cases where you would have to write unsafe code, you are making it explicit so that issues can easily be traced down to specific code blocks.
+
+### Memory safety
+
+Rust ensures memory safety at compile time using its innovative ownership mechanism and the borrow checker built into the compiler. The compiler just does not allow memory unsafe code unless it's explicitly marked as unsafe in an unsafe block or function. This static compile-time analysis eliminates many types of memory bugs, and with some more runtime checks, Rust guarantees memory safety.
+There is no concept of null at the language level. Instead, Rust provides the `Option` [enum](http://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/enums.html), which can be used to mark the presence or absence of a value, making the resulting code null safe and much easier to deal with, and you will never encounter null pointer exceptions in Rust.
+
+The ownership and borrowing mechanism make Rust one of the most memory-efficient languages while avoiding pitfalls with manual memory management and garbage collection. It has memory efficiency and speeds comparable to C/C++, and memory safety better than garbage collected languages like Java and Go.
+
+I wrote detailed articles about [memory management in different languages](https://deepu.tech/memory-management-in-programming/) in my personal blog, so check it out if you are interested in learning more about memory management in Java, Rust, JavaScript, and Go.
```

Comment:
```suggestion
I've written detailed articles about [memory management in different languages](https://deepu.tech/memory-management-in-programming/) in my personal blog, so check them out if you are interested in learning more about memory management in Java, Rust, JavaScript, and Go.
```
using past perfect gives describes that you wrote the articles over a period of time, rather than at one particular moment in the past.